components/sup/src/manager/service/terminator.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2019 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::spawned_future::SpawnedFuture;
+use crate::sys::{service, ShutdownMethod};
+use futures::sync::oneshot;
+use habitat_core::os::process::Pid;
+use std::thread;
+
+static LOGKEY: &'static str = "ST"; // "Service Terminator"
+
+/// Shut a service process down.
+///
+/// This is performed in a separate thread in order to prevent
+/// blocking the rest of the Supervisor.
+pub fn terminate_service(pid: Pid, service_group: String) -> SpawnedFuture<ShutdownMethod> {
+    let (tx, rx) = oneshot::channel();
+
+    let handle_result = thread::Builder::new()
+        .name(format!("{}-{}", LOGKEY, pid))
+        .spawn(move || {
+            outputln!(preamble service_group, "Terminating service (PID: {:?})", pid);
+            let shutdown = service::kill(pid);
+            match shutdown {
+                ShutdownMethod::AlreadyExited => {
+                    outputln!(preamble service_group, "Service already exited (PID: {:?})", pid);
+                }
+                ShutdownMethod::GracefulTermination => {
+                    outputln!(preamble service_group, "Service gracefully terminated (PID: {:?})", pid);
+                }
+                ShutdownMethod::Killed => {
+                    outputln!(preamble service_group, "Service killed (PID: {:?})", pid);
+                }
+            }
```

Comment:
I'm a sucker for deduplication and saving a few lines:
```rust
            let msg = match shutdown {
                ShutdownMethod::AlreadyExited => "Service already exited",
                ShutdownMethod::GracefulTermination => "Service gracefully terminated",
                ShutdownMethod::Killed => "Service killed",
            };
            outputln!(preamble service_group, "{} (PID: {:?})", msg, pid);
```
Or, go all the way and add
```rust
impl fmt::Display for ShutdownMethod {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let msg = match self {
            ShutdownMethod::AlreadyExited => "Service already exited",
            ShutdownMethod::GracefulTermination => "Service gracefully terminated",
            ShutdownMethod::Killed => "Service killed",
        };
        write!(f, "{}", msg)
    }
}
```
to `sup/src/sys.rs` and this becomes wonderfully concise:
```rust
            outputln!(preamble service_group, "{} (PID: {:?})", shutdown, pid);
```

Reply:
Oh, that is quite lovely. Stealing that now.